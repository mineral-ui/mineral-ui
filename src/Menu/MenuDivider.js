/**
 * Copyright 2017 CA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* @flow */
import React from 'react';
import { createStyledComponent } from '../styles';

type Props = Object;

export const componentTheme = (baseTheme: Object) => ({
  MenuDivider_borderColor: baseTheme.borderColor,
  MenuDivider_borderWidth: '1px',
  MenuDivider_margin: baseTheme.space_stack_sm,

  ...baseTheme
});

const Root = createStyledComponent(
  'div',
  props => {
    const theme = componentTheme(props.theme);

    return {
      backgroundColor: theme.MenuDivider_borderColor,
      height: theme.MenuDivider_borderWidth,
      margin: `${theme.MenuDivider_margin} 0`
    };
  },
  {
    displayName: 'MenuDivider'
  }
);

/**
 * MenuDividers separate [MenuItems](../menu-item) and [MenuGroups](../menu-group) to establish hierarchy in a [Menu](../menu) with several options.
 *
 * MenuDividers are not merely design elements to visually separate options.
 * Rather, the are a way to logically group choices.
 * Too many dividers will add unnecessary weight to your Menu.
 */
export default function MenuDivider(props: Props) {
  return <Root {...props} role="separator" />;
}
