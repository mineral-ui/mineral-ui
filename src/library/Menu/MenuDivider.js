/* @flow */
import React from 'react';
import { createStyledComponent } from '../styles';

type Props = Object;

export const componentTheme = (baseTheme: Object) => ({
  MenuDivider_borderColor: baseTheme.borderColor,
  MenuDivider_borderWidth: '1px',
  MenuDivider_margin: baseTheme.space_stack_sm,

  ...baseTheme
});

const Root = createStyledComponent(
  'div',
  props => {
    const theme = componentTheme(props.theme);

    return {
      backgroundColor: theme.MenuDivider_borderColor,
      height: theme.MenuDivider_borderWidth,
      margin: `${theme.MenuDivider_margin} 0`
    };
  },
  {
    displayName: 'MenuDivider'
  }
);

/**
 * MenuDividers separate [MenuItems](/components/menu-item) and [MenuGroups](/components/menu-group) to establish hierarchy in a [Menu](/components/menu) with several options.
 *
 * MenuDividers are not merely design elements to visually separate options.
 * Rather, the are a way to logically group choices.
 * Too many dividers will add unnecessary weight to your Menu.
 */
export default function MenuDivider(props: Props) {
  return <Root {...props} role="separator" />;
}
