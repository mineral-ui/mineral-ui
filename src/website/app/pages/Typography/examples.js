/* @flow */
import React from 'react';
import { createStyledComponent } from '../../../../library/styles';
import Text from '../../../../library/Text';

type TypeExample = {
  description: string,
  examples: Array<Object>
};

type TypeExamples = Array<TypeExample>;

const styles = {
  list: ({ theme }) => ({
    fontSize: theme.fontSize_ui,
    listStyle: 'square',
    paddingLeft: theme.space_inline_lg
  })
};

const List = createStyledComponent('ul', styles.list);
const examples: TypeExamples = [
  {
    description: `### Long-form Text

Default text and caption styles are intended for sections of content that
provide description and assistance. **Paragraph text** is primarily used for
content sections. **Caption Text** is used to highlight or offset content, such
as inline help or when captioning dashboard content.`,
    examples: [
      {
        content: <Text>Paragraph</Text>,
        value: '<p>, <li>',
        color: 'color_gray_100',
        font: 'Regular (14/0.875)'
      },
      {
        content: <Text appearance="prose">Paragraph (Prose)</Text>,
        value: '<p.prose>',
        color: 'color_gray_100',
        font: 'Regular (16/1)'
      },
      {
        content: <Text appearance="mouse">Paragraph (Mouse)</Text>,
        value: '<p.mouse>',
        color: 'color_gray_100',
        font: 'Regular (11/0.6875)'
      },
      {
        content: (
          <List>
            <li>Alpha</li>
            <li>Bravo</li>
            <li>Charlie</li>
          </List>
        ),
        // the design notes have <ol> here as well,
        // but list-style will override any numbers shown on <ol> list items
        value: '<ul>, <ol>',
        color: 'color_gray_100',
        font: 'Regular (14/0.875)'
      },
      {
        content: <Text element="h5">Caption - Strong</Text>,
        value: '<h5>',
        color: 'color_gray_100',
        font: 'Bold (14/0.875)'
      },
      {
        content: <Text element="h6">Caption - Light</Text>,
        value: '<h6>',
        color: 'color_gray_80',
        font: 'Regular (14/0.875)'
      }
    ]
  },
  {
    description: `### Display Text

Use **Display text** sparingly to make a bold statement, or draw attention to
different sections in an application view. Display text is for short titles and
section divisions—it’s not for long-form content.`,
    examples: [
      {
        content: <Text element="h1">Page Title</Text>,
        value: '<h1>',
        color: 'color_gray_100',
        font: 'ExtraBold (34/2.125)'
      },
      {
        content: <Text element="h2">Large Section Header</Text>,
        value: '<h2>',
        color: 'color_gray_80',
        font: 'Bold (28/1.75)'
      },
      {
        content: <Text element="h3">Medium Section Header</Text>,
        value: '<h3>',
        color: 'color_gray_80',
        font: 'Bold (22/1.375)'
      },
      {
        content: <Text element="h4">Small Section Header</Text>,
        value: '<h4>',
        color: 'color_gray_80',
        font: 'Bold (18/1.125)'
      }
    ]
  }
];

export default examples;
