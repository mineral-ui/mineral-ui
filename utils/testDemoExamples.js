/* @flow */
import React from 'react';
import { mount } from 'enzyme';
import { LiveProvider, LivePreview } from 'react-live';
import ThemeProvider from '../src/library/themes/ThemeProvider';

type Example = {
  title: React$Node,
  scope?: Object,
  id: string,
  source?: string
};
type Examples = Array<Example>;
type Options = {
  exclude?: Array<string> // Example id
};

export default function testDemoExamples(
  examples: Examples,
  options: Options = {}
) {
  if (options.exclude) {
    const exclusions = options.exclude || [];
    examples = examples.filter(example => !exclusions.includes(example.id));
  }

  return describe('demo examples', () => {
    examples.filter(({ scope, source }) => scope && source).map(example => {
      it(example.id, () => {
        const component = mount(
          <ThemeProvider>
            <LiveProvider
              code={example.source}
              scope={example.scope}
              mountStylesheet={false}>
              <LivePreview />
            </LiveProvider>
          </ThemeProvider>
        );
        expect(component).toMatchSnapshot();
      });
    });
  });
}
