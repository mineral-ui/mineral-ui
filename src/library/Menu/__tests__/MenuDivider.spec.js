/* @flow */
import React from 'react';
import { shallow } from 'enzyme';
import MenuDivider from '../MenuDivider';
import examples from '../../../website/app/demos/Menu/examples/MenuDivider';
import testDemoExamples from '../../../../utils/testDemoExamples';

function shallowMenuDivider(props = {}) {
  const menuDividerProps = {
    ...props
  };

  return shallow(<MenuDivider {...menuDividerProps} />);
}

describe('MenuDivider', () => {
  describe('renders', () => {
    it('root', () => {
      const menuDivider = shallowMenuDivider();

      expect(menuDivider.exists()).toEqual(true);
    });
  });

  testDemoExamples(examples);
});
