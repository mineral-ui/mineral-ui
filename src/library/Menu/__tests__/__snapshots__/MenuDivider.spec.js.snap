// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuDivider demo examples separating-items 1`] = `
.css-0,
[data-css-0] {
  overflow: hidden;
}

.css-0 > div,
[data-css-0] > div {
  background-color: #fff;
  float: left;
  margin: 0 0.5rem 0.5rem 0;
  width: 14em;
}

.css-1,
[data-css-1] {
  box-sizing: border-box;
  color: #333840;
  font-family: Open Sans, -apple-system, system-ui, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
  font-size: 16px;
  line-height: 1.25;
  outline: 0;
  -moz-osx-font-smoothing: auto;
  -webkit-font-smoothing: antialiased;
}

.css-1 *,
[data-css-1] *,
.css-1 *::before,
[data-css-1] *::before,
.css-1 *::after,
[data-css-1] *::after {
  box-sizing: inherit;
}

.css-2,
[data-css-2] {
  color: #333840;
  cursor: pointer;
  font-weight: 400;
  padding: 0.5em 1em;
}

.css-2:focus,
[data-css-2]:focus,
.css-2[data-simulate-focus],
[data-css-2][data-simulate-focus] {
  background-color: #f5f7fa;
  outline: 0;
}

.css-2:hover,
[data-css-2]:hover,
.css-2[data-simulate-hover],
[data-css-2][data-simulate-hover] {
  background-color: #f5f7fa;
}

.css-2:active,
[data-css-2]:active,
.css-2[data-simulate-active],
[data-css-2][data-simulate-active] {
  background-color: #ebeff5;
}

.css-2 [role="img"],
[data-css-2] [role="img"] {
  box-sizing: content-box;
  display: block;
  fill: #2e6fd9;
  flex: 0 0 auto;
  -webkit-flex: 0 0 auto;
}

.css-2 [role="img"]:first-child,
[data-css-2] [role="img"]:first-child {
  margin-right: 0.5em;
}

.css-2 [role="img"]:last-child,
[data-css-2] [role="img"]:last-child {
  margin-left: 0.5em;
}

.css-13,
[data-css-13] {
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  justify-content: space-between;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
}

.css-14,
[data-css-14] {
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  flex: 1 1 auto;
  flex-wrap: wrap;
  font-size: 0.875em;
  justify-content: space-between;
  padding: 0.21428571428571427em 0 0.2857142857142857em;
  white-space: normal;
  word-break: break-all;
  -webkit-flex: 1 1 auto;
  -webkit-box-lines: multiple;
  -webkit-flex-wrap: wrap;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
}

.css-15,
[data-css-15] {
  margin-right: 0.5714285714285714em;
  word-break: break-word;
}

.css-16,
[data-css-16] {
  color: #58606e;
  font-size: 0.75em;
  padding-top: 0.25em;
  word-break: break-word;
}

.css-17,
[data-css-17] {
  background-color: #c8d1e0;
  height: 1px;
  margin: 0.5em 0;
}

.css-18,
[data-css-18] {
  margin: 0.5em 0;
}

.css-18:first-child,
[data-css-18]:first-child,
.css-18 + .css-18,
[data-css-18] + [data-css-18] {
  margin-top: 0;
}

.css-18:last-child,
[data-css-18]:last-child,
.css-18[data-simulate-lastchild],
[data-css-18][data-simulate-lastchild] {
  margin-bottom: 0;
}

.css-24,
[data-css-24] {
  font-size: 0.75em;
  font-weight: 700;
  margin: 0;
  padding: 1.3333333333333333em 1.3333333333333333em 0.6666666666666666em;
}

<glamorous(div)>
  <div
    className="css-0"
  >
    <Menu>
      <Menu>
        <div
          className="css-1"
        >
          <MenuItem>
            <MenuItem
              tabIndex={0}
            >
              <div
                className="css-2"
                tabIndex={0}
              >
                <glamorous(span)>
                  <span
                    className="css-13"
                  >
                    <glamorous(span)>
                      <span
                        className="css-14"
                      >
                        <glamorous(span)>
                          <span
                            className="css-15"
                          >
                            Menu item
                          </span>
                        </glamorous(span)>
                        <glamorous(span)>
                          <span
                            className="css-16"
                          />
                        </glamorous(span)>
                      </span>
                    </glamorous(span)>
                  </span>
                </glamorous(span)>
              </div>
            </MenuItem>
          </MenuItem>
          <MenuDivider>
            <MenuDivider
              role="separator"
            >
              <div
                className="css-17"
                role="separator"
              />
            </MenuDivider>
          </MenuDivider>
          <MenuItem>
            <MenuItem
              tabIndex={0}
            >
              <div
                className="css-2"
                tabIndex={0}
              >
                <glamorous(span)>
                  <span
                    className="css-13"
                  >
                    <glamorous(span)>
                      <span
                        className="css-14"
                      >
                        <glamorous(span)>
                          <span
                            className="css-15"
                          >
                            Menu item
                          </span>
                        </glamorous(span)>
                        <glamorous(span)>
                          <span
                            className="css-16"
                          />
                        </glamorous(span)>
                      </span>
                    </glamorous(span)>
                  </span>
                </glamorous(span)>
              </div>
            </MenuItem>
          </MenuItem>
          <MenuItem>
            <MenuItem
              tabIndex={0}
            >
              <div
                className="css-2"
                tabIndex={0}
              >
                <glamorous(span)>
                  <span
                    className="css-13"
                  >
                    <glamorous(span)>
                      <span
                        className="css-14"
                      >
                        <glamorous(span)>
                          <span
                            className="css-15"
                          >
                            Menu item
                          </span>
                        </glamorous(span)>
                        <glamorous(span)>
                          <span
                            className="css-16"
                          />
                        </glamorous(span)>
                      </span>
                    </glamorous(span)>
                  </span>
                </glamorous(span)>
              </div>
            </MenuItem>
          </MenuItem>
          <MenuGroup
            title="Menu Group Title"
          >
            <MenuGroup>
              <div
                className="css-18"
              >
                <MenuGroupTitle>
                  <MenuGroupTitle>
                    <h3
                      className="css-24"
                    >
                      Menu Group Title
                    </h3>
                  </MenuGroupTitle>
                </MenuGroupTitle>
                <MenuItem>
                  <MenuItem
                    tabIndex={0}
                  >
                    <div
                      className="css-2"
                      tabIndex={0}
                    >
                      <glamorous(span)>
                        <span
                          className="css-13"
                        >
                          <glamorous(span)>
                            <span
                              className="css-14"
                            >
                              <glamorous(span)>
                                <span
                                  className="css-15"
                                >
                                  Menu item
                                </span>
                              </glamorous(span)>
                              <glamorous(span)>
                                <span
                                  className="css-16"
                                />
                              </glamorous(span)>
                            </span>
                          </glamorous(span)>
                        </span>
                      </glamorous(span)>
                    </div>
                  </MenuItem>
                </MenuItem>
                <MenuItem>
                  <MenuItem
                    tabIndex={0}
                  >
                    <div
                      className="css-2"
                      tabIndex={0}
                    >
                      <glamorous(span)>
                        <span
                          className="css-13"
                        >
                          <glamorous(span)>
                            <span
                              className="css-14"
                            >
                              <glamorous(span)>
                                <span
                                  className="css-15"
                                >
                                  Menu item
                                </span>
                              </glamorous(span)>
                              <glamorous(span)>
                                <span
                                  className="css-16"
                                />
                              </glamorous(span)>
                            </span>
                          </glamorous(span)>
                        </span>
                      </glamorous(span)>
                    </div>
                  </MenuItem>
                </MenuItem>
              </div>
            </MenuGroup>
          </MenuGroup>
          <MenuDivider>
            <MenuDivider
              role="separator"
            >
              <div
                className="css-17"
                role="separator"
              />
            </MenuDivider>
          </MenuDivider>
          <MenuGroup>
            <MenuGroup>
              <div
                className="css-18"
              >
                <MenuItem>
                  <MenuItem
                    tabIndex={0}
                  >
                    <div
                      className="css-2"
                      tabIndex={0}
                    >
                      <glamorous(span)>
                        <span
                          className="css-13"
                        >
                          <glamorous(span)>
                            <span
                              className="css-14"
                            >
                              <glamorous(span)>
                                <span
                                  className="css-15"
                                >
                                  Menu item
                                </span>
                              </glamorous(span)>
                              <glamorous(span)>
                                <span
                                  className="css-16"
                                />
                              </glamorous(span)>
                            </span>
                          </glamorous(span)>
                        </span>
                      </glamorous(span)>
                    </div>
                  </MenuItem>
                </MenuItem>
                <MenuItem>
                  <MenuItem
                    tabIndex={0}
                  >
                    <div
                      className="css-2"
                      tabIndex={0}
                    >
                      <glamorous(span)>
                        <span
                          className="css-13"
                        >
                          <glamorous(span)>
                            <span
                              className="css-14"
                            >
                              <glamorous(span)>
                                <span
                                  className="css-15"
                                >
                                  Menu item
                                </span>
                              </glamorous(span)>
                              <glamorous(span)>
                                <span
                                  className="css-16"
                                />
                              </glamorous(span)>
                            </span>
                          </glamorous(span)>
                        </span>
                      </glamorous(span)>
                    </div>
                  </MenuItem>
                </MenuItem>
              </div>
            </MenuGroup>
          </MenuGroup>
        </div>
      </Menu>
    </Menu>
  </div>
</glamorous(div)>
`;
